internal/engine/exit/conv.go

```diff
@@ -0,0 +1,254 @@
+		target.State.Waiting = &tiltruns.TargetStateWaiting{
+			Reason: pod.Status,
+		}
+	}
+
+	return target
+}
+
+func localServeTarget(mt *store.ManifestTarget, holds buildcontrol.HoldSet) *tiltruns.Target {
+	if mt.Manifest.LocalTarget().ServeCmd.Empty() {
+		// there is no runtime target
+		return nil
+	}
+
+	target := &tiltruns.Target{
+		Name:      fmt.Sprintf("%s:serve", mt.Manifest.Name.String()),
+		Resources: []string{mt.Manifest.Name.String()},
+		Type:      tiltruns.TargetTypeServer,
+	}
+
+	lrs := mt.State.LocalRuntimeState()
+	if runtimeErr := lrs.RuntimeStatusError(); runtimeErr != nil {
+		target.State.Terminated = &tiltruns.TargetStateTerminated{
+			StartTime:  metav1.NewMicroTime(lrs.StartTime),
+			FinishTime: metav1.NewMicroTime(lrs.FinishTime),
+			Error:      errToString(runtimeErr),
+		}
+	} else if lrs.PID != 0 {
+		target.State.Active = &tiltruns.TargetStateActive{
+			StartTime: metav1.NewMicroTime(lrs.StartTime),
+			Ready:     lrs.Ready,
+		}
+	} else {
+		target.State.Waiting = waitingFromHolds(mt.Manifest.Name, holds)
+	}
+
+	return target
+}
+
+func genericRuntimeTarget(mt *store.ManifestTarget, holds buildcontrol.HoldSet) *tiltruns.Target {
+	target := &tiltruns.Target{
+		Name:      fmt.Sprintf("%s:runtime", mt.Manifest.Name.String()),
+		Resources: []string{mt.Manifest.Name.String()},
+		Type:      tiltruns.TargetTypeServer,
+	}
+
+	// HACK: RuntimeState is not populated until engine starts builds in some cases; to avoid weird race conditions,
+	// 	it defaults to pending assuming the resource isn't _actually_ disabled on startup via auto_init=False
+	var runtimeStatus model.RuntimeStatus
+	if mt.State.RuntimeState != nil {
+		runtimeStatus = mt.State.RuntimeState.RuntimeStatus()
+	} else if mt.Manifest.TriggerMode.AutoInitial() {
+		runtimeStatus = model.RuntimeStatusPending
+	}
+
+	switch runtimeStatus {
+	case model.RuntimeStatusPending:
+		target.State.Waiting = waitingFromHolds(mt.Manifest.Name, holds)
+	case model.RuntimeStatusOK:
+		target.State.Active = &tiltruns.TargetStateActive{
+			StartTime: metav1.NewMicroTime(mt.State.LastSuccessfulDeployTime),
+			// generic resources have no readiness concept so they're just ready by default
+			// (this also applies to Docker Compose, since we don't support its health checks)
+			Ready: true,
+		}
+	case model.RuntimeStatusError:
+		errMsg := errToString(mt.State.RuntimeState.RuntimeStatusError())
+		if errMsg == "" {
+			errMsg = "Server target %q failed"
+		}
+		target.State.Terminated = &tiltruns.TargetStateTerminated{
+			Error: errMsg,
+		}
+	}
+
+	return target
+}
+
+func runtimeTarget(mt *store.ManifestTarget, holds buildcontrol.HoldSet) *tiltruns.Target {
+	if mt.Manifest.IsK8s() {
+		return k8sRuntimeTarget(mt)
+	} else if mt.Manifest.IsLocal() {
+		return localServeTarget(mt, holds)
+	} else {
+		return genericRuntimeTarget(mt, holds)
+	}
+}
+
+func buildTarget(mt *store.ManifestTarget, holds buildcontrol.HoldSet) *tiltruns.Target {
```

Comment:
Correct - I added an explanation in code, good call.

There's unfortunately no way to extract that out of the engine right now, so currently these are "aggregate"/virtual targets that can encompass multiple builds like in the examples you gave.

Hopefully, this can happen incrementally as part of other apiserver transitions (hand-waving on how it exactly gets consumed 😬), at which point this method, if it still exists, might return many targets rather than just one.

That said, the actual API schema accounts for this, hence exposing targets as the top-level with a `Resources []string` field, so that when you have many targets for a resource and/or a shared target across resources, it can be represented cleanly.

Reply:
cool, that makes sense-- i think a comment and otherwise punting for the future is AOK!